### ISSUE TITLE
Record failure on .NET (ok until 3.0.28)

### ISSUE THREAD

**Issue by StefanBertels** [louthy/language-ext#530]
I updated LanguageExt in a project and now get errors.

Version 3.0.28 works, Version 3.0.31 and 3.1.13 raise errors in my project.
I have a unit test showing the problem in my code but I could not reproduce it outside of my project. After copying the test and types into LanguageExt source code (test project) the test is green there (current master == 3.1.13). I'm confused why same test shows different behaviour.

My guess: it is something around 2f1e06. 

```
System.TypeInitializationException: The type initializer for 'LanguageExt.RecordType`1' threw an exception.
   at LanguageExt.Record`1.Equals(RECORDTYPE other)
   at MyProject.Actors.MyStaticClass.MyRecordClass.Equals(MyRecordClass rhs) in MyStaticClass.cs:line 488
   at Xunit.Sdk.AssertEqualityComparer`1.Equals(T x, T y)
   at Xunit.Assert.Equal[T](T expected, T actual, IEqualityComparer`1 comparer)
   at Xunit.Assert.Equal[T](T expected, T actual)
   at MyProject.Actors.MyTests.TestMyRecordClass() in MyTests.cs:line 22
System.InvalidOperationException: CompileToMethod cannot compile constant 'MyProject.Actors.MyStaticClass+MyRecordClass' because it is a non-trivial value, such as a live object. Instead, create an expression tree that can construct this value.
   at System.Linq.Expressions.Compiler.BoundConstants.EmitConstant(LambdaCompiler lc, Object value, Type type)
   at System.Linq.Expressions.Compiler.LambdaCompiler.EmitConstant(Object value, Type type)
   at System.Linq.Expressions.Compiler.LambdaCompiler.EmitConstantExpression(Expression expr)
   at System.Linq.Expressions.Compiler.LambdaCompiler.EmitExpression(Expression node, CompilationFlags flags)
   at System.Linq.Expressions.Compiler.LambdaCompiler.EmitBinaryExpression(Expression expr, CompilationFlags flags)
   at System.Linq.Expressions.Compiler.LambdaCompiler.EmitExpression(Expression node, CompilationFlags flags)
   at System.Linq.Expressions.Compiler.LambdaCompiler.EmitUnliftedAndAlso(BinaryExpression b)
   at System.Linq.Expressions.Compiler.LambdaCompiler.EmitAndAlsoBinaryExpression(Expression expr, CompilationFlags flags)
   at System.Linq.Expressions.Compiler.LambdaCompiler.EmitExpression(Expression node, CompilationFlags flags)
   at System.Linq.Expressions.Compiler.LambdaCompiler.EmitTypeBinaryExpression(Expression expr)
   at System.Linq.Expressions.Compiler.LambdaCompiler.EmitExpression(Expression node, CompilationFlags flags)
   at System.Linq.Expressions.Compiler.LambdaCompiler.EmitBranchComparison(Boolean branch, BinaryExpression node, Label label)
   at System.Linq.Expressions.Compiler.LambdaCompiler.EmitExpressionAndBranch(Boolean branchValue, Expression node, Label label)
   at System.Linq.Expressions.Compiler.LambdaCompiler.EmitUnliftedAndAlso(BinaryExpression b)
   at System.Linq.Expressions.Compiler.LambdaCompiler.EmitAndAlsoBinaryExpression(Expression expr, CompilationFlags flags)
   at System.Linq.Expressions.Compiler.LambdaCompiler.EmitExpression(Expression node, CompilationFlags flags)
   at System.Linq.Expressions.Compiler.LambdaCompiler.EmitUnliftedAndAlso(BinaryExpression b)
   at System.Linq.Expressions.Compiler.LambdaCompiler.EmitAndAlsoBinaryExpression(Expression expr, CompilationFlags flags)
   at System.Linq.Expressions.Compiler.LambdaCompiler.EmitExpression(Expression node, CompilationFlags flags)
   at System.Linq.Expressions.Compiler.LambdaCompiler.EmitUnliftedAndAlso(BinaryExpression b)
   at System.Linq.Expressions.Compiler.LambdaCompiler.EmitAndAlsoBinaryExpression(Expression expr, CompilationFlags flags)
   at System.Linq.Expressions.Compiler.LambdaCompiler.EmitExpression(Expression node, CompilationFlags flags)
   at System.Linq.Expressions.Compiler.LambdaCompiler.EmitUnliftedOrElse(BinaryExpression b)
   at System.Linq.Expressions.Compiler.LambdaCompiler.EmitOrElseBinaryExpression(Expression expr, CompilationFlags flags)
   at System.Linq.Expressions.Compiler.LambdaCompiler.EmitExpression(Expression node, CompilationFlags flags)
   at System.Linq.Expressions.Compiler.LambdaCompiler.EmitLambdaBody(CompilerScope parent, Boolean inlined, CompilationFlags flags)
   at System.Linq.Expressions.Compiler.LambdaCompiler.EmitLambdaBody()
   at System.Linq.Expressions.Compiler.LambdaCompiler.Compile(LambdaExpression lambda, MethodBuilder method, DebugInfoGenerator debugInfoGenerator)
   at System.Linq.Expressions.LambdaExpression.CompileToMethodInternal(MethodBuilder method, DebugInfoGenerator debugInfoGenerator)
   at LanguageExt.IL.Compile[F](LambdaExpression lambda, String typeName, String methodName)
   at LanguageExt.IL.Equals[A](Boolean includeBase)
   at LanguageExt.RecordType`1..cctor()
```

```
public static class MyStaticClass
{
	internal class MyRecordClass : Record<MyRecordClass>
	{
		public MyRecordClass(MyEvent @event)
		{
			Name = @event.Name;
			BirthDay = @event.BirthDay;
		}

		public MyRecordClass(string name, DateTime birthDay)
		{
			Name = name;
			BirthDay = birthDay;
		}

		public DateTime BirthDay { get; }

		[OptOutOfEq, OptOutOfHashCode, OptOutOfOrd]
		//[Eq(typeof(EqStringOrdinalIgnoreCase))]
		public string Name { get; }

		public override bool Equals(MyRecordClass rhs) =>
			base.Equals(rhs) && EqStringOrdinalIgnoreCase.Inst.Equals(rhs.Name, Name);
	}
}
```

```
[Fact]
public static void BirthdayTest()
{
	var key1 = new MyStaticClass.MyRecordClass("john", DateTime.Parse("2018-09-19"));
	var key2 = new MyStaticClass.MyRecordClass("John", DateTime.Parse("2018-09-19"));
	var key3 = new MyStaticClass.MyRecordClass("Jane", DateTime.Parse("2018-09-19"));
	Assert.Equal(key1, key2);
	Assert.NotEqual(key3, key2);
	Assert.Equal(0, key1.CompareTo(key2));
	Assert.Equal(key1.GetHashCode(), key2.GetHashCode());
}
```

I tested the new attributes / `NonStructural` -- same problem.

Any idea how to reproduce this in LanguageExt / what the difference might be?

**Comment by StefanBertels** [comment]
Maybe related: https://stackoverflow.com/questions/35903724/compiletomethod-cannot-compile-constant-some-value-because-it-is-a-non-trivial

**Comment by StefanBertels** [comment]
I stripped down my project to only a few lines. Test throws runtime error here. If I downgrade to LanguageExt 3.0.28 everything works fine (and test green).

I can even remove my custom Equals method (#247) and it still fails.

my.csproj
```
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>net462</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="LanguageExt.Core" Version="3.0.31" />
    <PackageReference Include="xunit" Version="2.4.1" PrivateAssets="All" />
    <PackageReference Include="xunit.runner.visualstudio" Version="2.4.1" PrivateAssets="all" />
  </ItemGroup>
</Project>
```

SomeStaticClass.cs
```
using LanguageExt;
using LanguageExt.ClassInstances;
using Xunit;

public static class SomeStaticClass
{
    [Fact]
    static void TestMyRecord()
    {
        var key1 = new MyRecord("a");
        var key2 = new MyRecord("A");
        var key3 = new MyRecord("b");
        Assert.Equal(key1, key2);
        Assert.NotEqual(key3, key2);
        Assert.Equal(0, key1.CompareTo(key2));
        Assert.Equal(key1.GetHashCode(), key2.GetHashCode());
    }
}

internal class MyRecord : Record<MyRecord>
{
    public MyRecord(string text)
    {
        Text = text;
    }

    [OptOutOfEq, OptOutOfHashCode, OptOutOfOrd] 
    public string Text { get; }

    public override bool Equals(MyRecord rhs) =>
        base.Equals(rhs) && EqStringOrdinalIgnoreCase.Inst.Equals(rhs.Text, Text);
}
```

**Comment by StefanBertels** [comment]
I just could reproduce it in LanguageExt.Tests! If you modify `LanguageExt.Tests.csproj` this way, some tests fail: `<TargetFrameworks>net462</TargetFrameworks>`

**Comment by louthy** [comment]
@StefanBertels I am mega busy at the moment, I don't suppose you've narrowed down the change/diff that caused the problem?

**Comment by bender2k14** [comment]
Oh, I just realized that this issue is also blocking me from upgrading Language Ext.  When I tried to upgrade some months back, many of my tests failed, but I never had the chance to debug why.

We are targeting .Net 4.6.1, and I just confirmed that my tests pass using version 3.0.28 but fail with version 3.0.31 by throwing an exception when trying to access some property backing fields of a `Record<>` type.
